### PR TITLE
connections: do not modify selector during iteration

### LINKED ISSUE #312: timed_out_connections - dictionary changed size during iteration
❓ **I'm submitting a ...**
- [X] 🐞 bug report
- [ ] 🐣 feature request
- [ ] ❓ question about the decisions made in the repository

🐞 **Describe the bug. What is the current behavior?**
Traceback in the logs:
```python-traceback
Traceback (most recent call last):
  File "/opt/lib/python3.8/site-packages/cheroot/server.py", line 1795, in serve
    self.tick()
  File "/opt/lib/python3.8/site-packages/cheroot/server.py", line 2030, in tick
    self.connections.expire()
  File "/opt/lib/python3.8/site-packages/cheroot/connections.py", line 107, in expire
    for sock_fd, conn in timed_out_connections:
  File "/opt/lib/python3.8/site-packages/cheroot/connections.py", line 102, in <genexpr>
    timed_out_connections = (
  File "/opt/lib/python3.8/_collections_abc.py", line 743, in __iter__
RuntimeError: dictionary changed size during iteration
```

❓ **What is the motivation / use case for changing the behavior?**

💡 **To Reproduce**

💡 **Expected behavior**

📋 **Details**

📋 **Environment**

- Cheroot version: 8.4.4
- CherryPy version: 18.6.0 (if applicable)
- Python version: 3.8.X
- OS: NAS
- Browser: all

📋 **Additional context**

### PR DESCRIPTION
**Pull request by liamstask**
❓ **What kind of change does this PR introduce?**
  - [x] 🐞 bug fix
  - [ ] 🐣 feature
  - [ ] 📋 docs update
  - [ ] 📋 tests/coverage improvement
  - [ ] 📋 refactoring
  - [ ] 💥 other

📋 **What is the related issue number (starting with `#`)**

Fixes #312

❓ **What is the current behavior?** (You can also link to an open issue here)

The selector map is modified during iteration, raising an error.

❓ **What is the new behavior (if this is a feature change)?**

Create a temporary list of the items to remove, then remove them after iterating the selector map.

📋 **Other information**:

📋 **Checklist**:

  - [x] I think the code is well written
  - [x] I wrote [good commit messages][1]
  - [x] I have [squashed related commits together][2] after the changes have been approved
  - [ ] Unit tests for the changes exist
  - [ ] Integration tests for the changes exist (if applicable)
  - [x] I used the same coding conventions as the rest of the project
  - [x] The new code doesn't generate linter offenses
  - [x] Documentation reflects the changes
  - [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

---
This change is [Reviewable](https://reviewable.io/reviews/cherrypy/cheroot/313)

### CODE REVIEW

File: cheroot/connections.py [hunk 1]
```diff
@@ -101,7 +102,8 @@ def expire(self):
         threshold = time.time() - self.server.timeout
         timed_out_connections = (
             (sock_fd, conn)
-            for _, (_, sock_fd, _, conn) in self._selector.get_map().items()
+            for _, (_, sock_fd, _, conn)
+            in copy.copy(self._selector.get_map()).items()
```
Comment (webknjaz) [inline cheroot/connections.py:106]: @liamstask why is a deep copy necessary? Won't a shallow copy do the job?

Comment (webknjaz) [reply cheroot/connections.py:106]: Oh, I just saw https://github.com/cherrypy/cheroot/pull/311#discussion_r470056286. Well, maybe it's alright, then. Ah, it's not even a deep copy. I think it shoulds fine, then.

**Comment by webknjaz** [comment]
@Safihre mind confirming that this change fixes your issue?

**Comment by webknjaz** [comment]
@Safihre also, it'd be great if we could come up with a stable reproducer that could be integrated into the test suite as a regression test.

File: cheroot/connections.py (hunk 1 above)
Comment (liamstask) [reply cheroot/connections.py:106]: yes - https://docs.python.org/3.8/library/copy.html#copy.copy specifically mentions this is a shallow copy.

**Comment by Safihre** [comment]
It did not work:
```
Traceback (most recent call last):
  File "C:\Users\safv\Documents\GitHub\sabnzbd3\venv\lib\site-packages\cheroot\server.py", line 1797, in serve
    self.tick()
  File "C:\Users\safv\Documents\GitHub\sabnzbd3\venv\lib\site-packages\cheroot\server.py", line 2032, in tick
    self.connections.expire()
  File "C:\Users\safv\Documents\GitHub\sabnzbd3\venv\lib\site-packages\cheroot\connections.py", line 109, in expire
    for sock_fd, conn in timed_out_connections:
  File "C:\Users\safv\Documents\GitHub\sabnzbd3\venv\lib\site-packages\cheroot\connections.py", line 103, in <genexpr>
    timed_out_connections = (
  File "C:\Users\safv\AppData\Local\Programs\Python\Python38\lib\_collections_abc.py", line 743, in __iter__
    for key in self._mapping:
RuntimeError: dictionary changed size during iteration
```
I double checked the files, and it does have the `copy.copy` changes. 
The reproducer is simply clicking around a bit and wait for the periodic check. Will see if I can make `cheroot` only reproducer.

**Comment by Safihre** [comment]
I can reproduce with:
```python
from cheroot import wsgi
def my_crazy_app(environ, start_response):
    status = '200 OK'
    response_headers = [('Content-type','text/plain')]
    start_response(status, response_headers)
    return [b'Hello world!']

addr = '0.0.0.0', 8070
server = wsgi.Server(addr, my_crazy_app)
server.start()
```
I open `localhost:8070` in the browser, wait a few seconds and then it happens.

**Comment by webknjaz** [comment]
@Safihre is it windows-specific?

**Comment by webknjaz** [comment]
@Safihre it'd be great to have something in the form of a test that pytest would run. But since the CIs currently cannot reproduce this, I think there may be some extra preconditions to it.

**Comment by OneCDOnly** [comment]
> @Safihre is it windows-specific?

@webknjaz no, seeing the same think in Linux:
```
2020-08-13 12:04:08,114::ERROR::[_cplogging:213] [13/Aug/2020:12:04:08] ENGINE Error in HTTPServer.tick
Traceback (most recent call last):
  File "/opt/lib/python3.8/site-packages/cheroot/server.py", line 1795, in serve
    self.tick()
  File "/opt/lib/python3.8/site-packages/cheroot/server.py", line 2030, in tick
    self.connections.expire()
  File "/opt/lib/python3.8/site-packages/cheroot/connections.py", line 107, in expire
    for sock_fd, conn in timed_out_connections:
  File "/opt/lib/python3.8/site-packages/cheroot/connections.py", line 102, in <genexpr>
    timed_out_connections = (
  File "/opt/lib/python3.8/_collections_abc.py", line 743, in __iter__
RuntimeError: dictionary changed size during iteration
```

**Comment by liamstask** [comment]
if you're able to test again, i've pushed a change that does not copy the map, but instead creates a temporary list of the items to be removed, such that the selector is not modified during iteration.

**Comment by Safihre** [comment]
Tested, and this one works! 👍 
Have not been able to verify if it was Windows only today.

**Comment by webknjaz** [comment]
@liamstask weird, it seems like a band-aid because the list creation still iterates over the items... It just does it faster.

**Comment by webknjaz** [comment]
Oh, you've also found where you modify the map. That's better. So I guess it's reasonable to revert it to use a generator comprehension, then.

File: cheroot/connections.py [hunk 2]
```diff
@@ -99,11 +99,12 @@ def expire(self):
...
         threshold = time.time() - self.server.timeout
-        timed_out_connections = (
+        timed_out_connections = [
             (sock_fd, conn)
-            for _, (_, sock_fd, _, conn) in self._selector.get_map().items()
+            for _, (_, sock_fd, _, conn)
+            in self._selector.get_map().items()
             if conn != self.server and conn.last_used < threshold
-        )
+        ]
```
Comment (webknjaz) [inline cheroot/connections.py:107]: Can we please revert this change? Since the code below stops modifying the map, it should be no longer necessary.

Comment (webknjaz) [reply cheroot/connections.py:107]: Oh, it's a separate function. I take it back, then.

**Comment by webknjaz** [comment]
@liamstask do you think you could come up with a regression test for this?

**Comment by liamstask** [comment]
i cannot commit to a specific timeline, but i can try to take a look

**Comment by Safihre** [comment]
I also noticed a new traceback with the new code when exiting the server with some requests still going:
```
2020-08-14 14:44:01,587::INFO::[_cplogging:213] [14/Aug/2020:14:44:01] ENGINE Bus STOPPING
2020-08-14 14:44:01,801::INFO::[_cplogging:213] [14/Aug/2020:14:44:01] ENGINE HTTP Server cherrypy._cpwsgi_server.CPWSGIServer(('127.0.0.1', 8081)) shut down
2020-08-14 14:44:01,801::INFO::[_cplogging:213] [14/Aug/2020:14:44:01] ENGINE Bus STOPPED
2020-08-14 14:44:01,801::INFO::[_cplogging:213] [14/Aug/2020:14:44:01] ENGINE Bus EXITING
2020-08-14 14:44:01,801::INFO::[_cplogging:213] [14/Aug/2020:14:44:01] ENGINE Bus EXITED
2020-08-14 14:44:01,805::ERROR::[_cplogging:213] [14/Aug/2020:14:44:01] ENGINE TypeError("object of type 'NoneType' has no len()")
Traceback (most recent call last):
  File "c:\users\safv\documents\github\sabnzbd3\venv\lib\site-packages\cheroot\server.py", line 1283, in communicate
    req.respond()
  File "c:\users\safv\documents\github\sabnzbd3\venv\lib\site-packages\cheroot\server.py", line 1073, in respond
    self.server.gateway(self).respond()
  File "c:\users\safv\documents\github\sabnzbd3\venv\lib\site-packages\cheroot\wsgi.py", line 145, in respond
    self.write(chunk)
  File "c:\users\safv\documents\github\sabnzbd3\venv\lib\site-packages\cheroot\wsgi.py", line 229, in write
    self.req.ensure_headers_sent()
  File "c:\users\safv\documents\github\sabnzbd3\venv\lib\site-packages\cheroot\server.py", line 1120, in ensure_headers_sent
    self.send_headers()
  File "c:\users\safv\documents\github\sabnzbd3\venv\lib\site-packages\cheroot\server.py", line 1165, in send_headers
    can_keep = self.server.connections.can_add_keepalive_connection
  File "c:\users\safv\documents\github\sabnzbd3\venv\lib\site-packages\cheroot\connections.py", line 292, in can_add_keepalive_connection
    return ka_limit is None or self._num_connections < ka_limit
  File "c:\users\safv\documents\github\sabnzbd3\venv\lib\site-packages\cheroot\connections.py", line 286, in _num_connections
    return len(self._readable_conns) + len(self._selector.get_map()) - 1
TypeError: object of type 'NoneType' has no len()
```

**Comment by liamstask** [comment]
that appears to be due to the server using `connections` after it has already been closed - pushed a small change to address (would also benefit from a regression test case).

**Comment by Safihre** [comment]
@liamstask Tested the new version and it seems to not show any tracebacks.

**Comment by Safihre** [comment]
Just got a new one with the latest code, during regular serving:
```
Traceback (most recent call last):
  File "C:\Users\safv\Documents\GitHub\sabnzbd3\venv\lib\site-packages\cheroot\server.py", line 1797, in serve
    self.tick()
  File "C:\Users\safv\Documents\GitHub\sabnzbd3\venv\lib\site-packages\cheroot\server.py", line 2032, in tick
    self.connections.expire()
  File "C:\Users\safv\Documents\GitHub\sabnzbd3\venv\lib\site-packages\cheroot\connections.py", line 107, in expire
    for sock_fd, conn in timed_out_connections:
  File "C:\Users\safv\Documents\GitHub\sabnzbd3\venv\lib\site-packages\cheroot\connections.py", line 102, in <genexpr>
    timed_out_connections = (
  File "C:\Users\safv\AppData\Local\Programs\Python\Python38\lib\_collections_abc.py", line 743, in __iter__
    for key in self._mapping:
RuntimeError: dictionary changed size during iteration
```

**Comment by webknjaz** [comment]
@Safihre this traceback looks like https://github.com/cherrypy/cheroot/pull/313#issuecomment-673671850 but there's no generator expression in that place anymore. Are you sure you've checked out the latest incarnation of this branch?

P.S. Pro tip: you can mark your code blocks with `python-traceback` and they'll show up highlighted.

**Comment by webknjaz** [comment]
@liamstask I think I'll wait until there's some tests added. No rush here: I'll be on vacation this week and will probably be unable to do proper reviews anyway.

**Comment by Safihre** [comment]
@webknjaz You seem to be right! Seems my pip-install went wrong.

File: cheroot/server.py [hunk 3]
```diff
@@ -1162,7 +1162,10 @@ def send_headers(self):  # noqa: C901  # FIXME
         # Override the decision to not close the connection if the connection
         # manager doesn't have space for it.
         if not self.close_connection:
-            can_keep = self.server.connections.can_add_keepalive_connection
+            can_keep = (
+                self.server.ready
```
Comment (webknjaz) [inline cheroot/server.py:1166]: Maybe embed this into `connections.can_add_keepalive_connection` property?

**Comment by webknjaz** [comment]
@liamstask should I expect you to add the test this weekend? If not, I'll just merge this right away and hope that the test gets added in some follow-up PR.

**Comment by cyraxjoe** [comment]
PR with the regression tests https://github.com/cherrypy/cheroot/pull/316

File: cheroot/server.py (hunk 3 above)
Comment (liamstask) [reply cheroot/server.py]: in c94eaabb400f885a93aa9bf79a7153a6d8581d41 i instead made connections a 'private' member of server, since it should never be accessed after `server.stop()` has been called.

**Comment by liamstask** [comment]
@cyraxjoe thanks, that makes sense to me.

@webknjaz i added one more commit to address one of your previous comments, related to protecting access of `server.connections` - i generalized that a bit and made it internal to server altogether. if that looks ok to you, i can squash all these commits prior to merging.

**Comment by webknjaz** [comment]
It's alright to keep the commits separate. In fact, I was about to merge this yesterday but got distracted.
